Additional modules for interacting with RabbitMQ

jaconza opened the pull request:
Modules to allow users to add and remove queues, bindings and
exchanges using the RabbitMQ management plugin's http based API.

Since the default rabbitmq control script does not allow for adding queues, exchanges or bindings we opted to make use of the management plugin's http based api.

Commit:
Additional modules for interacting with RabbitMQ

Modules to allow developers to add and remove queues, bindings and
exchanges using the RabbitMQ management plugin's http based API.

jimbydamonk commented:
Very nice. I was about to start working on this.

jimbydamonk commented:
I can't seem to get your examples to work when using arguments.
I copied your example, and get the following:

---

 ERROR: Syntax Error while loading YAML script, test.yml
Note: The error may actually appear before this position: line 10, column 55

```
              queue=test_queue
              arguments='{"x-dead-letter-exchange": "dead"}'
                                                  ^
```

This one looks easy to fix.  It seems that there is a value started
with a quote, and the YAML parser is expecting to see the line ended
with the same kind of quote.  For instance:

```
when: "ok" in result.stdout
```

Could be written as:

   when: '"ok" in result.stdout'

or equivalently:

   when: "'ok' in result.stdout"

Commit:
Merge branch 'devel' of github.com:ansible/ansible into module_rabbitmq_api

Commit:
Fix the examples for the rabbitmq api modules.

The examples were using an incorrect host argument and the arguments
argument was failing in the yaml whenever it has a space in it.

jaconza commented:
@jimbydamonk It seem like yaml does not like the space in the arguments string. I have update the example to remove this space. 

The following yields an error:

```
arguments='{"x-dead-letter-exchange": "dead"}'
```

But removing the space solves the error:

```
arguments='{"x-dead-letter-exchange":"dead"}'
```

While doing this I also realised that the host argument was wrong in most of the example. This to have been update. 

Here is the example playbook I used for testing:

``` yaml
- hosts: local
  user: root 
  tasks:
    - rabbitmq_queue: host='localhost'
                      user=guest
                      password=guest
                      vhost=/
                      queue=test_queue
                      arguments='{"x-dead-letter-exchange":"dead"}'

    - rabbitmq_exchange: host='localhost'
                         user=guest
                         password=guest
                         vhost=/
                         name=test_exchange
                         type=direct
                         durable=no
                         internal=no
                         auto_delete=no
                         arguments='{"some-additional-arg":true}'

    - rabbitmq_binding: host='localhost'
                        user='guest'
                        password='guest'
                        vhost='/'
                        source='test_exchange'
                        destination='test_queue'
                        binding_type='queue'
                        routing_key='test_routing_key'
```

jimbydamonk commented:
@jaconel That worked. Removing the space works. 

Now on to the next error. If I store the args in a variable it seems to be hitting a json parse error. I assume this is something to do with the way jinja is messing with the var. 
Here is my play book:

```
---
- hosts: localhost
  vars:
    product_queues: 
      - { name: 'ha.reject', arguments: '{"x-dead-letter-exchange":"dead"}' }
  tasks:
    - name: Debug
      debug: msg=" args = {{ item.arguments }}"
      with_items: product_queues 

    - name: Test rabbit no var 
      rabbitmq_queue: user=guest password=guest vhost=/ queue=test_queue 
        arguments='{"x-dead-letter-exchange":"dead"}'    
      ignore_errors: yes

    - name: Test rabbit single var  
      rabbitmq_queue: user=guest password=guest vhost=/ queue=test_queue 
        arguments=" {{ product_queues[0].arguments }}"   
      ignore_errors: yes

    - name: Test Rabbit in a loop 
      rabbitmq_queue: user=guest password=guest vhost=/ queue=test_queue 
        arguments="{{ item.arguments }}"
      with_items: product_queues
```

I put the out put in a gist https://gist.github.com/jimbydamonk/5700cc06b8db20743bdb

You can see that the tasks with a variable die with:
`ValueError: Expecting property name: line 1 column 3 (char 2)`

jimbydamonk commented:
Oh, and I hit a bit with the binding:
on line 198 of rabbit_binding you have self.bindings instead of just bindings. 
Removing self. makes it work.

Commit:
Referencing incorrect propery in rabbitm_binding module.

On line 198 of rabbit_binding the code was referencing self.bindings
instead of just the local scoped bindings.

jaconza commented:
So managed to fix the scope error in rabbitmq_binding. Thanks for that. 

Still looking into the "args in variable" error. It would appear that Jinja is stripping the `"` from the string, thus causing the json parser in the module to fail. Will confirm shortly. Have already tried using the `safe` filter, no love.

jaconza commented:
Ok, gave it a quick test. It seems to be that Jinja is stripping the quotes from the strings. Test can be seen here: https://gist.github.com/jaconel/1dfddf21f2467d520ee5

Commit:
Changed rabbitmq api modules to rather take complex object type for args

The 'arguments' parameter for the queue and binding modules are no
longer json strings, but rather complex object types. Updated the
examples to show some different usage scenarios.

jaconza commented:
Broke away from passing the 'arguments' parameter in as a json string. Instead used a complex argument and altered examples.

Commit:
Updated rabbitmq api module's version added propery.

mpdehaan commented:
Hi!

Thanks very much for your interest in Ansible.  It sincerely means a lot to us. 

On September 26, 2014, due to enormous levels of contribution to the project Ansible decided to reorganize module repos, making it easier
for developers to work on the project and for us to more easily manage new contributions and tickets.

We split modules from the main project off into two repos, http://github.com/ansible/ansible-modules-core and http://github.com/ansible/ansible-modules-extras

If you still would like this pull request merged, we will need your help making this target the new repo.  If you do not take any action, this
pull request unfortunately cannot be applied.

We apologize that we are not able to make this transition happen seamlessly, though this is a one-time change and your help is greatly appreciated -- 
this will greatly improve velocity going forward.

Both sets of modules will ship with Ansible, though they'll receive slightly different ticket handling.

To locate where a module lives between 'core' and 'extras'
- Find the module at http://docs.ansible.com/list_of_all_modules.html
- Open the documentation page for that module
- If the bottom of the docs say "This is an extras module", submit your ticket to https://github.com/ansible/ansible-modules-extras
- Otherwise, submit your pull request to update the existing module to https://github.com/ansible/ansible-modules-core
- Note that python modules in ansible now also end in ".py" and this extension is required for new contributions.
- action_plugins (modules with server side components) still live in the main repo.  If your pull request touches both, which should be
   exceedingly rare, submit two new pull requests and make sure to mention the links to each other in the comments.

Otherwise, if this is a new module:
- Submit your pull request to add a module to https://github.com/ansible/ansible-modules-extras

It may be possible to re-patriate your pull requests automatically, one user-submitted approach for advanced git users
has been suggested at https://gist.github.com/willthames/afbaaab0c9681ed45619

Additionally, should you need more help with this, you can ask questions on:
- the development mailing list: https://groups.google.com/forum/#!forum/ansible-devel

Thanks you very much!
